lib/segment/src/utils/mem.rs

```diff
@@ -0,0 +1,189 @@
+    }
+}
+
+pub fn assert_available_memory_during_segment_load(
+    mem: &Mem,
+    data_size_bytes: u64,
+    available_memory_percent: u64,
+) -> OperationResult<()> {
+    let total_memory_bytes = mem.total_memory_bytes();
+    let available_memory_bytes = mem.available_memory_bytes();
+
+    let available_memory_after_data_load_bytes =
+        available_memory_bytes.saturating_sub(data_size_bytes);
+
+    let required_available_memory_bytes = total_memory_bytes / 100 * available_memory_percent;
+
+    // Create segment only if at least 15% of RAM available (to prevent OOM on load)
+    if available_memory_after_data_load_bytes >= required_available_memory_bytes {
+        Ok(())
+    } else {
+        Err(OperationError::service_error(format!(
+            "less than {available_memory_percent}% of RAM available \
+             ({available_memory_after_data_load_bytes} bytes out of {total_memory_bytes} bytes) \
+             while loading DB segment, segment load aborted to prevent OOM"
+        )))
+    }
+}
+
+#[cfg(target_os = "linux")]
+mod cgroups_mem {
+    use cgroups_rs::{hierarchies, memory, Cgroup};
+    use procfs::process::Process;
+
+    #[derive(Clone, Debug)]
+    pub struct CgroupsMem {
+        mem_controller: memory::MemController,
+        memory_limit_bytes: Option<u64>,
+        used_memory_bytes: u64,
+    }
+
+    impl CgroupsMem {
+        pub fn new() -> Option<Self> {
+            let memory_cgroup_path = match get_current_process_memory_cgroup_path() {
+                Ok(memory_cgroup_path) => memory_cgroup_path?,
+                Err(err) => {
+                    log::error!(
+                        "Failed to query current process info \
+                         while initializing CgroupsMem: {err}"
+                    );
+
+                    return None;
+                }
+            };
+
+            let cgroup = Cgroup::load(
+                hierarchies::auto(),
+                memory_cgroup_path.trim_start_matches('/'),
+            );
+
+            let mut mem = Self {
+                mem_controller: cgroup.controller_of::<memory::MemController>()?.clone(),
+                memory_limit_bytes: None,
+                used_memory_bytes: 0,
+            };
+
+            mem.refresh();
+
+            Some(mem)
+        }
+
+        pub fn refresh(&mut self) {
+            let stat = self.mem_controller.memory_stat();
+            self.memory_limit_bytes = stat.limit_in_bytes.try_into().ok();
+            self.used_memory_bytes = stat.usage_in_bytes;
+        }
+
+        pub fn memory_limit_bytes(&self) -> Option<u64> {
+            self.memory_limit_bytes
+        }
+
+        pub fn used_memory_bytes(&self) -> u64 {
+            self.used_memory_bytes
+        }
+    }
+
+    fn get_current_process_memory_cgroup_path() -> procfs::ProcResult<Option<String>> {
+        let process = Process::myself()?;
+        let cgroups = process.cgroups()?;
+
+        for cgroup in cgroups {
```

Comment:
Thanks for the review, @timvisee! I think it makes more sense to check for v2 first, as Docker (being the biggest slowpoke here) got support for (completely switched to?) cgroups v2, like, 2 years ago, and, generally, cgroups v2 are available for ~8 years now, so v1 is a legacy (and so, an afterthought) at this point.

Reply:
Makes sense!

I figured that the v1 check may be a bit more reliable because we test for a specific controller rather than for it being empty.

But on second thought I agree, check v2 first.

Reply:
I figured v1 _have to_ have _some_ controller. (I may be wrong, though. I focused on and tested v2.)